### PR TITLE
service/apt: Implement Deliver Args

### PR DESCRIPTION
**Pull request by zhaowenlan1779**
This implements deliver args for application jumping.

Fixes part of #3105. You can now launch the games, but most throw fatal errors upon start.

Please refer to the commit messages for details.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/5611)

### CODE REVIEW

File: src/core/hle/service/apt/applet_manager.h [hunk 1]
```diff
@@ -242,6 +271,7 @@ class AppletManager : public std::enable_shared_from_this<AppletManager> {
     };
 
     ApplicationJumpParameters app_jump_parameters{};
+    boost::optional<DeliverArg> deliver_arg{};
```
Comment (B3n30) [inline src/core/hle/service/apt/applet_manager.h:274]: std::optional ?

File: src/core/hle/service/apt/apt.cpp [hunk 2]
```diff
@@ -449,19 +453,32 @@ void Module::APTInterface::PrepareToDoApplicationJump(Kernel::HLERequestContext&
...
+        hmac_size = max_hmac_size;
+    }
 
-    LOG_WARNING(Service_APT, "(STUBBED) called param_size={:08X}, hmac_size={:08X}", param_size,
-                hmac_size);
+    auto param = rp.PopStaticBuffer();
+    auto hmac = rp.PopStaticBuffer();
 
-    // TODO(Subv): Set the delivery parameters before starting the new application.
+    LOG_WARNING(Service_APT, "called param_size={:08X}, hmac_size={:08X}", param_size, hmac_size);
```
Comment (B3n30) [inline src/core/hle/service/apt/apt.cpp:477]: I think this shouldn't be a warning log level

File: src/core/core.cpp [hunk 3]
```diff
@@ -561,9 +563,20 @@ void System::Reset() {
     // reloading.
     // TODO: Properly implement the reset
 
+    // Since the system is completely reinitialized, we'll have to store the deliver arg manually.
+    boost::optional<Service::APT::AppletManager::DeliverArg> deliver_arg;
```
Comment (BreadFish64) [inline src/core/core.cpp:567]:
```suggestion
    std::optional<Service::APT::AppletManager::DeliverArg> deliver_arg;
```

File: src/core/hle/service/apt/apt.cpp [hunk 4]
```diff
@@ -479,6 +496,25 @@ void Module::APTInterface::GetProgramIdOnApplicationJump(Kernel::HLERequestConte
     rb.Push(static_cast<u8>(parameters.next_media_type));
 }
 
+void Module::APTInterface::ReceiveDeliverArg(Kernel::HLERequestContext& ctx) {
+    IPC::RequestParser rp(ctx, 0x35, 2, 4); // 0x00350080
+    const auto param_size = rp.Pop<u32>();
+    const auto hmac_size = rp.Pop<u32>();
+
+    LOG_WARNING(Service_APT, "called param_size={:08X}, hmac_size={:08X}", param_size, hmac_size);
```
Comment (B3n30) [inline src/core/hle/service/apt/apt.cpp:504]: Shouldn't be LOG_WARNING probably DEBUG or INFO

File: src/core/hle/service/apt/apt.cpp [hunk 5]
```diff
@@ -479,6 +496,25 @@ void Module::APTInterface::GetProgramIdOnApplicationJump(Kernel::HLERequestConte
...
+
+    auto arg = apt->applet_manager->ReceiveDeliverArg().value_or(AppletManager::DeliverArg{});
+    arg.param.resize(param_size);
```
Comment (B3n30) [inline src/core/hle/service/apt/apt.cpp:507]: there is no check for param_size?

File: src/core/hle/service/apt/applet_manager.cpp [hunk 6]
```diff
@@ -498,23 +498,31 @@ ResultCode AppletManager::PrepareToDoApplicationJump(u64 title_id, FS::MediaType
...
+    if (app_jump_parameters.flags != ApplicationJumpFlags::UseCurrentParameters) {
+        // The source program ID is not updated when using flags 0x2.
+        arg.source_program_id = application_slot.title_id;
+    }
+
     application_slot.Reset();
 
-    // TODO(Subv): Set the delivery parameters.
+    // Set the delivery parameters.
+    deliver_arg = arg;
```
Comment (BreadFish64) [inline src/core/hle/service/apt/applet_manager.cpp:525]:
```suggestion
    deliver_arg = std::move(arg);
```

File: src/core/hle/service/apt/apt.cpp [hunk 7]
```diff
@@ -790,15 +826,36 @@ void Module::APTInterface::GetStartupArgument(Kernel::HLERequestContext& ctx) {
...
     }
 
-    std::vector<u8> parameter(parameter_size);
+    std::vector<u8> param;
+    bool exists = false;
+
+    if (auto arg = apt->applet_manager->ReceiveDeliverArg()) {
+        param = std::move(arg->param);
+
+        // TODO: This is a complete guess based on observations. It is unknown how the OtherMedia
```
Comment (B3n30) [inline src/core/hle/service/apt/apt.cpp:835]: This should probably get logged (warning) then

File: src/core/hle/service/apt/applet_manager.h [hunk 8]
```diff
@@ -158,7 +160,30 @@ class AppletManager : public std::enable_shared_from_this<AppletManager> {
...
+        }
+        friend class boost::serialization::access;
+    };
+
+    ResultCode DoApplicationJump(DeliverArg arg);
+
+    boost::optional<DeliverArg> ReceiveDeliverArg() const {
+        return deliver_arg;
+    }
+    void SetDeliverArg(boost::optional<DeliverArg> arg) {
```
Comment (BreadFish64) [inline src/core/hle/service/apt/applet_manager.h:184]:
```suggestion
    void SetDeliverArg(std::optional<DeliverArg> arg) {
```

File: src/core/hle/service/apt/applet_manager.h (hunk 1 above)
Comment (BreadFish64) [inline src/core/hle/service/apt/applet_manager.h:274]:
```suggestion
    std::optional<DeliverArg> deliver_arg{};
```

**Review by BreadFish64 — COMMENTED** [review]
Looks okay to me for the most part.
I do think we've stopped using `boost::optional` in favor of `std::optional`.

**Comment by zhaowenlan1779** [comment]
boost::optional was used because of this comment

https://github.com/citra-emu/citra/blob/d36bfcaec37f9af8fed0678b2bf33557e849329f/src/core/hle/service/apt/applet_manager.h#L226-L227

File: src/core/hle/service/apt/apt.cpp (hunk 5 above)
Comment (zhaowenlan1779) [reply src/core/hle/service/apt/apt.cpp:507]: It has been verified by HW tests that the entire buffer was cleared even when the actual parameter was of a smaller size.

**Comment by B3n30** [comment]
removed from canary because of conflicts with master. Needs a rebase

**Comment by tywald** [comment]
A user has reported that this change caused Mii Maker to crash.

**Comment by Dragios** [comment]
```
Assertion Failed!: iter != fastmem_patch_info.end()
Debug <Critical> core/hle/kernel/hle_ipc.cpp:operator():260: Assertion Failed!
Static buffer data is too big
```
For Mii Maker Crash

**Pull request merged**